presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.geospatial;
+
+import com.esri.core.geometry.Geometry;
+import com.esri.core.geometry.GeometryEngine;
+import com.esri.core.geometry.OperatorImportFromESRIShape;
+import com.esri.core.geometry.SpatialReference;
+import com.esri.core.geometry.ogc.OGCGeometry;
+import com.facebook.presto.spi.PrestoException;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.airlift.slice.Slices.EMPTY_SLICE;
+import static java.util.Objects.requireNonNull;
+
+public final class GeometryUtils
+{
+    public static final String POINT = "Point";
+    public static final String LINE_STRING = "LineString";
+    public static final String POLYGON = "Polygon";
+    public static final String MULTI_POINT = "MultiPoint";
+    public static final String MULTI_LINE_STRING = "MultiLineString";
+    public static final String MULTI_POLYGON = "MultiPolygon";
+    public static final int WKID_UNKNOWN = 0;
+
+    private static final int SIZE_WKID = 4;
+
+    private GeometryUtils() {}
+
+    public static OGCGeometry deserialize(Slice shape)
+    {
+        requireNonNull(shape, "shape is null");
+        int wkid = shape.getInt(0);
+        ByteBuffer shapeBuffer = getShapeBuffer(shape);
+        if (shapeBuffer.limit() < SIZE_WKID || shapeBuffer.getInt(0) == Geometry.Type.Unknown.value()) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Corrupted bytes for Geometry");
+        }
+
+        SpatialReference spatialReference = null;
+        if (wkid != WKID_UNKNOWN) {
+            spatialReference = SpatialReference.create(wkid);
+        }
+        Geometry esriGeom = OperatorImportFromESRIShape.local().execute(0, Geometry.Type.Unknown, shapeBuffer);
+        OGCGeometry createdGeom = OGCGeometry.createFromEsriGeometry(esriGeom, spatialReference);
+        return createdGeom;
+    }
+
+    public static Slice serialize(OGCGeometry input)
+    {
+        int wkid = input.SRID();
+        Geometry geometry = input.getEsriGeometry();
+
+        if (geometry == null) {
```

Comment:
When input is an instance of GeometryCollection, the geometry will be null. Therefore serializing GeometryCollection will always produce an empty slice. Is this intentional?

--- from com.esri.core.geometry.ogc.OGCConcreteGeometryCollection

public Geometry getEsriGeometry() {
        return null;
    }

    public GeometryCursor getEsriGeometryCursor() {
        return new OGCConcreteGeometryCollection.GeometryCursorOGC(this.geometries);
    }

Reply:
yes, I did not find a good way to serialize GeometryCollection. What do you think?

Reply:
@zhenxiao , at first I thought that it might be OK to not support GeometryCollection by disallowing GeometryCollection in GeometryFromText. However, the problem is that GeometryCollection may be created as a result of some of the operations. For example, ST_Intersection may return a collection:

ST_Intersection(ST_GeometryFromText('POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))'), ST_GeometryFromText('LINESTRING (-1 1, 1 -1, 1 2)'))

returns a point and a linestring: GEOMETRYCOLLECTION (POINT (0 0), LINESTRING (1 0, 1 1))

GeometryCollection can be serialized as an array of geometries. ArrayBlockBuilder may be handy. One option is for serialize to always produce an array of geometries and for deserialize to create GEOMETRYCOLLECTION when array size > 1 and GEOMETRYCOLLECTION EMPTY when array is empty. What do you think?